sdk/core/azure-core-http-httpurlconnection/src/main/java/com/azure/core/http/httpurlconnection/HttpUrlConnectionAsyncClient.java

```diff
@@ -0,0 +1,353 @@
+                    + httpRequest.getHttpMethod()));
+            }
+        }
+        return requestSendMono;
+    }
+
+    /**
+     * Synchronously sends the content of an HttpRequest via an HttpUrlConnection instance.
+     *
+     * @param httpRequest The HTTP Request being sent
+     * @param progressReporter A reporter for the progress of the request
+     * @param connection The HttpURLConnection that is being sent to
+     * @return This method does not return any value
+     */
+    private Void sendBodySync(HttpRequest httpRequest, ProgressReporter progressReporter, HttpURLConnection connection) {
+        Void requestSendMono = null;
+
+        switch (httpRequest.getHttpMethod()) {
+            case POST:
+            case PUT:
+            case DELETE: {
+                connection.setDoOutput(true);
+                BinaryData binaryBodyData = httpRequest.getBodyAsBinaryData();
+
+                if (binaryBodyData != null) {
+                    byte[] bytes = binaryBodyData.toBytes();
+                    if (progressReporter != null) {
+                        progressReporter.reportProgress(bytes.length);
+                    }
+                    try (DataOutputStream os = new DataOutputStream(new BufferedOutputStream(connection.getOutputStream()))) {
+                        os.write(bytes);
+                        os.flush();
+                    } catch (IOException e) {
+                        throw LOGGER.logExceptionAsError(new RuntimeException(e));
+                    }
+                }
+            }
+            case GET:
+            case HEAD:
+            case OPTIONS:
+            case TRACE:
+            case CONNECT: {
+                break;
+            }
+            default: {
+                throw LOGGER.logExceptionAsError(new IllegalStateException("Unknown HTTP Method:"
+                    + httpRequest.getHttpMethod()));
+            }
+        }
+        return requestSendMono;
+    }
+
+    /**
+     * Receive the response from the remote server
+     *
+     * @param httpRequest The HTTP Request being sent
+     * @param connection The HttpURLConnection being sent to
+     * @return A HttpResponse object
+     */
+    private HttpResponse receiveResponse(HttpRequest httpRequest, HttpURLConnection connection) {
+        try {
+            int responseCode = connection.getResponseCode();
+
+            HttpHeaders responseHeaders = new HttpHeaders();
+            for (Map.Entry<String, List<String>> entry : connection.getHeaderFields().entrySet()) {
+                if (entry.getKey() != null) {
+                    List<String> values = new ArrayList<>();
+                    entry.getValue().forEach(v -> values.add(0, v));
+                    for (String headerValue : values) {
+                        responseHeaders.add(HttpHeaderName.fromString(entry.getKey()), headerValue);
+                    }
```

Comment:
You can sanity check me (as I just typed this in GitHub and didn't compile / test it, but is the code you have equivalent to the following:
```suggestion
                if (entry.getKey() != null) {
                    final HttpHeaderName headerName = HttpHeaderName.fromString(entry.getKey());
                    entry.getValue().forEach(headerValue -> responseHeaders.add(headerName, headerValue));
```